src/lib/registration/registration_codes_loader.rb

```diff
@@ -0,0 +1,100 @@
+# encoding: utf-8
+
+# ------------------------------------------------------------------------------
+# Copyright (c) 2016 SUSE, LLC. All Rights Reserved.
+#
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, contact Novell, Inc.
+#
+# To contact Novell about this file by physical or electronic mail, you may find
+# current contact information at www.novell.com.
+# ------------------------------------------------------------------------------
+
+require "tempfile"
+
+require "yast"
+require "transfer/file_from_url"
+
+Yast.import "XML"
+
+module Registration
+  # Aids Registration::Storage::RegCodes in loading the codes
+  # from external storage so that the user does not have to type them
+  # each time when doing multiple installations. FATE#316796
+  module RegCodesLoader
+    include Yast::I18n # missing in yast2-update
+    include Yast::Transfer::FileFromUrl
+
+    REGCODES_NAME_HANDLERS = {
+      "regcodes.xml" => :reg_codes_from_xml,
+      "regcodes.txt" => :reg_codes_from_txt
+    }
+
+    # @return [Hash{String => String},nil]
+    def reg_codes_from_usb_stick
+      REGCODES_NAME_HANDLERS.each do |name, handler|
+        with_tempfile(name) do |path|
+          if get_file_from_url(scheme: "usb", host: "", urlpath: "/#{name}",
+                               localfile: path,
+                               urltok: {}, destdir: "")
+            codes = send(handler, path)
+            return codes if codes
+          end
+        end
+      end
+      nil
+    end
+
+    # @param template for tempfile name
+    # @yieldparam actual file name
+    def with_tempfile(pattern, &block)
+      tempfile = Tempfile.new(pattern)
+      block.call(tempfile.path)
+    ensure
+      tempfile.unlink
+    end
+
+    # @param filename [String]
+    # @return [Hash{String => String},nil]
+    def reg_codes_from_xml(filename)
+      return nil unless File.readable?(filename)
+      xml_hash = Yast::XML.XMLToYCPFile(filename)
+      parse_xml(xml_hash)
+    end
+
+    # @param xml_hash [Hash] as used in AY and returned by Yast::XML.XMLToYCPFile
+    # @return [Hash{String => String},nil]
+    def parse_xml(xml_hash)
+      suse_register = xml_hash.fetch("suse_register", {})
+      addons = suse_register.fetch("addons", [])
```

Comment:
In the end reading suse_register/reg_code is not implemented, and base product regcode is read from the addon section. Reading the other element would IMO complicate the code too much for little benefit. We can add later it if we get feedback.